README.md
# standard-output

## A. Description
Standard exit points in every functions.
There are 3 types of exit point:
- Success Response
- Expected Error
- Unexpected Error


## B. How to use

### 1. Installation
```
npm install --save @dobert/standard-output
```

### 2. Importing
```
const {
    SuccessResponse,
    ExpectedError,
    UnexpectedError,
    OutputTypes,
    StatusCodes
} = require('@dobert/standard_output');
```

### 3. SET the result

#### SuccessResponse
```
const result = new SuccessResponse({
    message: 'Success to ....',
    detail: StatusCodes.REQUEST_SUCCESS,
    data: {} // data you want to binding on the response
});
```

Parameters description:
- message is a simple message that you want give to the user and it must be easy to understand even by non-technical person
- detail for success response usually is `DATA_CREATED` or `REQUEST_SUCCESS`
- data is a information that you want to bind to the response

#### ExpectedError
```
const error = new ExpectedError();
error.additionalInfo({
    message: 'Failed to ...',
    detail: StatusCodes.UNAUTHORIZED
});
```
Parameters description:
- message is a simple message that you want give to the user and it must be easy to understand even by non-technical person
- detail for expected error usually besides `DATA_CREATED` and `REQUEST_SUCCESS`

#### UnexpectedError
```
const error = new UnexpectedError(err);
```
Parameters description:
- err is from error we caught in `try catch` block


### 4. GET the processed result

```
result.getStandardResult();
```

this will return object like this
```
{
    code: 200,
    json: {
        message: 'Success to get data',
        data: {}
    }
}
```

## C. Update the code SOP
1. Change the branch
   - You must following this branch flow
     
     `new_branch` > `master` > `release`
  
   - Your `new_branch` name must start with one of these:
     - breaking-change/...
     - feature/...
     - bugfixing/...

2. Update the version at `package.json`, you can refer to this [versioning standard](https://semver.org/)

3. Update the `CHANGELOG.md`


## D. How to publish the npm

1. Login to the npm
   ```
   npm adduser
   ```
2. publish the npm
   ```
   npm publish --access public
   ```

NB:
- You MUST publish to npmjs only from branch release
- If this doesn't work, please refer to [npmjs documentation](https://www.npmjs.com/)


CHANGELOG.md
# Changelog

### version 2.1.4
- fixing typo alllow to allow

### version 2.1.3
- update Access-Control-Allow-Methods to all

### version 2.1.2
- bugfixing, typo in JSON.stringify

### version 2.1.1
- bugfixing, unexpected identifier

### version 2.1.0
- add headers to handle cors

### version 2.0.2
- change type error to response_type in response
- remove error_description in expected error

### version 2.0.1
- add type error in response

### version 2.0.0
- change to support AWS Lambda

### version 1.0.4
- remove unused file and update status code 404

### version 1.0.3
- add response code 404 (wrong updated file)

### version 1.0.2
- Bugfixing: wrong import file name

### version 1.0.1
- Rename some files to make it easier to read

### version 1.0.0
- Add initial standard output
  - success response
  - expected error
  - unexpected error
- Add initial response code (based on HTTP status code)


src/expected_error.js
const OutputTypes = require('./enums/output_types');

class ExpectedError extends Error {
    constructor () {
        super();

        this.type = OutputTypes.EXPECTED_ERROR;
        this.status;
        this.message;
    }

    additionalInfo(data) {
        this.status = data.status;
        this.message = data.message;
    }

    getStandardResult() {
        return {
            statusCode: this.status.http_code,
            headers: {
                "Access-Control-Allow-Headers" : "Content-Type",
                "Access-Control-Allow-Origin" : "*",
                "Access-Control-Allow-Methods" : "*",
            },
            body: JSON.stringify({
                response_type: this.type,
                message: this.message
            })
        }
    }
}

module.exports = ExpectedError;


src/unexpected_error.js
const OutputTypes = require('./enums/output_types');
const StatusCodes = require('./enums/status_codes');

class UnexpectedError extends Error {
    constructor (error) {
        super(error);
        this.type = OutputTypes.UNEXPECTED_ERROR;
        this.error = error;
    }

    getStandardResult() {
        return {
            statusCode: StatusCodes.INTERNAL_SERVER_ERROR.http_code,
            headers: {
                "Access-Control-Allow-Headers" : "Content-Type",
                "Access-Control-Allow-Origin" : "*",
                "Access-Control-Allow-Methods" : "*",
            },
            body: JSON.stringify({
                response_type: this.type,
                message: 'Something unexpected happened',
                error_description: this.error
            })
        }
    }
}

module.exports = UnexpectedError;


src/enums/status_codes.js
module.exports = {
    'REQUEST_SUCCESS': {
        http_code: 200,
        message: 'Success'
    },
    'DATA_CREATED': {
        http_code: 201,
        message: 'New resource is created in the collection'
    },
    'REQUEST_ACCEPTED': {
        http_code: 202,
        message: 'Request has been accepted for processing, but the processing has not been completed'
    },
    'NO_CONTENT': {
        http_code: 204,
        message: 'Data not found'
    },
    'BAD_REQUEST': {
        http_code: 400,
        message: 'Invalid request message parameters'
    },
    'UNAUTHORIZED': {
        http_code: 401,
        message: 'Can not authorize the credentials'
    },
    'FORBIDDEN': {
        http_code: 403,
        message: 'User does not have the necessary permissions for the resource'
    },
    'REQUEST_TIMEOUT': {
        http_code: 408,
        message: 'The server timed out waiting for the request'
    },
    'DATA_CONFLICT': {
        http_code: 409,
        message: 'Conflict in the current state of the resource'
    },
    'UNSUPPORTED_MEDIA_TYPE': {
        http_code: 415,
        message: 'The request entity has a media type which the server or resource does not support'
    },
    'INTERNAL_SERVER_ERROR': {
        http_code: 500,
        message: 'Unexpected error happened (generic error)'
    },
    'NOT_IMPLEMENTED': {
        http_code: 501,
        message: 'This feature is not available yet'
    }
}


src/enums/response_codes.js
module.exports = {
    'REQUEST_SUCCESS': {
        http_code: 200,
        message: 'Success'
    },
    'DATA_CREATED': {
        http_code: 201,
        message: 'New resource is created in the collection'
    },
    'REQUEST_ACCEPTED': {
        http_code: 202,
        message: 'Request has been accepted for processing, but the processing has not been completed'
    },
    'NO_CONTENT': {
        http_code: 204,
        message: 'Data not found'
    },
    'BAD_REQUEST': {
        http_code: 400,
        message: 'Invalid request message parameters'
    },
    'UNAUTHORIZED': {
        http_code: 401,
        message: 'Can not authorize the credentials'
    },
    'FORBIDDEN': {
        http_code: 403,
        message: 'User does not have the necessary permissions for the resource'
    },
    'DATA_NOT_FOUND': {
        http_code: 404,
        message: 'Data not found'
    },
    'REQUEST_TIMEOUT': {
        http_code: 408,
        message: 'The server timed out waiting for the request'
    },
    'DATA_CONFLICT': {
        http_code: 409,
        message: 'Conflict in the current state of the resource'
    },
    'UNSUPPORTED_MEDIA_TYPE': {
        http_code: 415,
        message: 'The request entity has a media type which the server or resource does not support'
    },
    'INTERNAL_SERVER_ERROR': {
        http_code: 500,
        message: 'Unexpected error happened (generic error)'
    },
    'NOT_IMPLEMENTED': {
        http_code: 501,
        message: 'This feature is not available yet'
    }
}
